### PR TITLE
Issue #5320 - using jetty-websocket-httpclient.xml within webapp Jetty 10

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #5320**

Make similar changes from PR #5374 for Jetty 10.

- re-enable the distribution tests for websocket client within webapp.
- fix the `XmlHttpClientProvider` so that it uses the HttpClient's classloader
- add the `websocket-jetty-client.jar` to jetty-home/lib/websocket/. This is so it can be optionally exposed to the webapp if you wish to use it as a  provided dependency.

### CODE REVIEW

File: jetty-home/pom.xml [hunk 1]
```diff
@@ -655,6 +655,11 @@
       <artifactId>websocket-jetty-server</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty.websocket</groupId>
+      <artifactId>websocket-jetty-client</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
Comment (sbordet) [inline jetty-home/pom.xml:662]: Why is this added to `jetty-home`? 
I think we always want `websocket-jetty-client` to be in webapps.

File: jetty-websocket/websocket-jetty-server/src/main/config/modules/websocket-jetty.mod [hunk 2]
```diff
@@ -18,6 +18,7 @@ lib/websocket/websocket-util-${jetty.version}.jar
 lib/websocket/websocket-util-server-${jetty.version}.jar
 lib/websocket/websocket-jetty-api-${jetty.version}.jar
 lib/websocket/websocket-jetty-common-${jetty.version}.jar
+lib/websocket/websocket-jetty-client-${jetty.version}.jar
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-server/src/main/config/modules/websocket-jetty.mod]: Don't think this is necessary?

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java [hunk 3]
```diff
@@ -389,11 +387,12 @@ public void testWebsocketClientInWebappProvidedByServer(String scheme) throws Ex
...
             .build();
 
+        String module = "https".equals(scheme) ? "test-keystore," + scheme : scheme;
         String[] args1 = {
             "--create-startd",
             "--approve-all-licenses",
-            "--add-to-start=resources,server,webapp,deploy,jsp,jmx,servlet,servlets,websocket,test-keystore," + scheme
-        };
+            "--add-to-start=resources,server,webapp,deploy,jsp,jmx,servlet,servlets,websocket," + module,
+            };
```
Comment (sbordet) [inline tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:395]: Fix indentation.

**Review by sbordet — CHANGES_REQUESTED** [review]
I don't think `jetty-home` or POM dependencies should depend on `websocket-jetty-client`.
If you added those dependencies just for testing, then we should find another way to test without adding the dependencies to the server modules.

File: jetty-home/pom.xml (hunk 1 above)
Comment (gregw) [reply jetty-home/pom.xml:662]: Do we?  Isn't the point of the module for the server to provide it?
Sure we want to support it in webapps as well, but always?

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java [hunk 4]
```diff
@@ -406,6 +405,9 @@ public void testWebsocketClientInWebappProvidedByServer(String scheme) throws Ex
             String[] args2 = {
                 "jetty.http.port=" + port,
                 "jetty.ssl.port=" + port,
+                // We need to expose the websocket client classes to the webapp for this to work.
+                "jetty.webapp.addServerClasses+=,-org.eclipse.jetty.websocket.client.",
+                "jetty.webapp.addSystemClasses+=,+org.eclipse.jetty.websocket.client.",
```
Comment (gregw) [inline tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java]: If the server is to provide this, then any exposing should be done in a Configuration... off by default, but that is added to a webapp if we want it to see the client.

File: jetty-home/pom.xml (hunk 1 above)
Comment (sbordet) [reply jetty-home/pom.xml:662]: Yes, I think we always want the Jetty specific WebSocket client to be in webapps.
No classloader issues, no issues when deploying (e.g. for testing) in other containers, etc.
There is no point adding it to `jetty-home`, but then cannot be seen by webapps and the server does not use it.

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java (hunk 4 above)
Comment (sbordet) [reply tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java]: Do I need to write a class? 

I thought a Jetty module should be enough, and the test mimics what should be in the `*.mod` file if I want the Jetty WebSocket client classes be provided by the server.

File: jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
+
+[description]
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod:3]: Remove these 2 lines, the link is now obsolete with the new Jetty 10 docs.

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import org.eclipse.jetty.webapp.WebAppConfiguration;
+import org.eclipse.jetty.webapp.WebInfConfiguration;
+import org.eclipse.jetty.webapp.WebXmlConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Websocket Configuration</p>
+ * <p>This configuration configures the WebAppContext server/system classes to
+ * be able to see the org.eclipse.jetty.websocket package.
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: Use javadoc's `{@code ...}` to wrap package names, and fix the package names because we expose `*.api` and `*.client` only, not `*.websocket`.

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    public boolean isAvailable()
+    {
+        return isAvailable("org.eclipse.jetty.websocket.client.WebSocketClient");
+    }
+
+    private boolean isAvailable(String classname)
+    {
+        try
+        {
+            return Loader.loadClass(classname) != null;
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: I'm not sure this is right.

It will try the context classloader, which is the webapp classloader, so if there is `WebSocketClient` in `WEB-INF/lib` (e.g. by mistake), it will return `true`, but we don't want that.

I think you want to test if the server has this class available, so perhaps `JettyWebSocketClientConfiguration.class.getClassLoader().loadClass(...)` should be used instead.

File: jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java [hunk 8]
```diff
@@ -68,7 +68,7 @@ else if (isAvailable("org.eclipse.jetty.annotations.AnnotationConfiguration"))
     @Override
     public boolean isAvailable()
     {
-        return isAvailable("org.eclipse.jetty.websocket.common.JettyWebSocketFrame");
+        return isAvailable("org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer");
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java]: See comment about using `Loader.loadClass()`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod (hunk 5 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod:3]: These lines are present on every single .mod file I can find. Maybe the change to fix all these should all be done in a separate PR?

File: jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+Expose the Jetty WebSocket Client classes to deployed web applications.
+
+[tags]
+websocket
+
+[depend]
+client
+annotations
+
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod:12]: Should this module depend on the jetty server websocket impl?    Once we expose the API aren't we committed to supplying the server side?

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+ */
+public class JettyWebSocketClientConfiguration extends AbstractConfiguration
+{
+    private static final Logger LOG = LoggerFactory.getLogger(JettyWebSocketClientConfiguration.class);
+
+    public JettyWebSocketClientConfiguration()
+    {
+        addDependencies(WebXmlConfiguration.class, MetaInfConfiguration.class, WebInfConfiguration.class, FragmentConfiguration.class);
+
+        if (isAvailable("org.eclipse.jetty.osgi.annotations.AnnotationConfiguration"))
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: I don't think you need to do this availability check. Just add the dependency as a string and it will be ordered if it is there and ignored if it is not.

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java (hunk 10 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: But the logic is only adding the osgi one if it is available, otherwise only adding the standard one. Wouldn't it have a different behaviour if we add both.

File: jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod (hunk 9 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-client/src/main/config/modules/websocket-jetty-client.mod:12]: We're not committed to supplying the server side by having the client. The websocket client can be used without any server dependencies.

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java (hunk 10 above)
Comment (gregw) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: But I don't think you need to do that even.  The OSGi configuration replaces the non OSGI configuration, so it should inherit all of it's dependencies and orderings.  You can't list out all the infinite possible extensions of org.eclipse.jetty.annotations.AnnotationConfiguration here, so don't try.

Comment (gregw) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: hmmm for some reason the OSGi AnnotationConfiguration does not "replace" the normal AnnotationConfiguration???
Might have to check with @janbartel why?
But ultimately here you should just depend on org.eclipse.jetty.annotations.AnnotationConfiguration and the replacement mechanism should make the sorting work correctly

Comment (gregw) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: Perhaps I just forgot to add it when I did the conversion.   Can you try adding to the OSGi AnnotationConfiguration:
```java

    @Override
    public Class<? extends Configuration> replaces()
    {
        return org.eclipse.jetty.annotations.AnnotationConfiguration.class;
    }
```

Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: I overrided the `replaces()` method in the osgi AnnotationConfiguration now. It passed the tests even without that, even after I made the changes you asked for removing the `isAvailable` checks.

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+ * </p>
+ */
+public class JettyWebSocketClientConfiguration extends AbstractConfiguration
+{
+    private static final Logger LOG = LoggerFactory.getLogger(JettyWebSocketClientConfiguration.class);
+
+    public JettyWebSocketClientConfiguration()
+    {
+        addDependencies(WebXmlConfiguration.class, MetaInfConfiguration.class, WebInfConfiguration.class, FragmentConfiguration.class);
+        addDependents("org.eclipse.jetty.osgi.annotations.AnnotationConfiguration", WebAppConfiguration.class.getName());
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: You don't need this line, only the next one!   Look at how replaces works.

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        hide("org.eclipse.jetty.client.impl.");
+        hide("org.eclipse.jetty.client.config.");
+    }
+
+    @Override
+    public boolean isAvailable()
+    {
+        try
+        {
+            ClassLoader classLoader = JettyWebSocketClientConfiguration.class.getClassLoader();
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: Are these two classes in the same jar?  If so, you kind of can just return true in this method (or not override).
This method is more for things like CDI and JSP, where we need to check for third party jars being available

File: jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java [hunk 13]
```diff
@@ -48,39 +48,29 @@
     public JettyWebSocketConfiguration()
     {
         addDependencies(WebXmlConfiguration.class, MetaInfConfiguration.class, WebInfConfiguration.class, FragmentConfiguration.class);
+        addDependents("org.eclipse.jetty.osgi.annotations.AnnotationConfiguration", WebAppConfiguration.class.getName());
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java]: Don't need this line.
```suggestion
```

File: jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java [hunk 14]
```diff
@@ -48,39 +48,29 @@
...
-        return isAvailable("org.eclipse.jetty.websocket.common.JettyWebSocketFrame");
-    }
-
-    private boolean isAvailable(String classname)
     {
         try
         {
-            return Loader.loadClass(classname) != null;
+            ClassLoader classLoader = JettyWebSocketConfiguration.class.getClassLoader();
+            return classLoader.loadClass("org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer") != null;
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/config/JettyWebSocketConfiguration.java]: Ditto - maybe just return true here.
Feel free to improve javadoc on isAvailable if you think it wasn't made clear

**Review by gregw — CHANGES_REQUESTED** [review]
better... but ....

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Websocket Configuration</p>
+ * <p>This configuration configures the WebAppContext server/system classes to
+ * be able to see the {@code org.eclipse.jetty.websocket.client} package.
+ * This class is defined in the webapp package, as it implements the {@link Configuration} interface,
+ * which is unknown to the websocket package.  However, the corresponding {@link ServiceLoader}
+ * resource is defined in the websocket package, so that this configuration only be
+ * loaded if the jetty-websocket jars are on the classpath.
```
Comment (sbordet) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/config/JettyWebSocketClientConfiguration.java]: I think the javadoc needs to be fixed as this class is not defined anymore in the "webapp package".

**Review by sbordet — APPROVED** [review]
LGTM but please fix the javadocs I refer to in the comment.
